core/lib/libtomcrypt/sm2.c

```diff
@@ -0,0 +1,632 @@
+
+	C1 = ltc_ecc_new_point();
+	if (!C1) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	res = sm2_bytes_to_point(C1, &ltc_key.dp, src, src_len, &C1_len);
+	if (res)
+		goto out;
+
+	/* Step B2: S = [h]C1 */
+
+	if (ltc_key.dp.cofactor != 1) {
+		S = ltc_ecc_new_point();
+		if (!S) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			goto out;
+		}
+
+		ltc_res = mp_init_multi(&h, NULL);
+		if (ltc_res != CRYPT_OK)
+			return TEE_ERROR_OUT_OF_MEMORY;
+
+		ltc_res = mp_set_int(h, ltc_key.dp.cofactor);
+		if (ltc_res != CRYPT_OK) {
+			res = TEE_ERROR_BAD_STATE;
+			goto out;
+		}
+
+		ltc_res = ltc_ecc_mulmod(h, C1, S, ltc_key.dp.A,
+					 ltc_key.dp.prime, 1);
+		if (ltc_res != CRYPT_OK) {
+			res = TEE_ERROR_BAD_STATE;
+			goto out;
+		}
+
+		ltc_res = ltc_ecc_is_point_at_infinity(S, ltc_key.dp.prime,
+						       &inf);
+	} else {
+		ltc_res = ltc_ecc_is_point_at_infinity(C1, ltc_key.dp.prime,
+						       &inf);
+	}
+	if (inf) {
+		res = TEE_ERROR_BAD_STATE;
+		goto out;
+	}
+
+	/* Step B3: (x2, y2) = [dB]C1 */
+
+	x2y2p = ltc_ecc_new_point();
+	if (!x2y2p) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	ltc_res = ltc_ecc_mulmod(ltc_key.k, C1, x2y2p, ltc_key.dp.A,
+				 ltc_key.dp.prime, 1);
+	if (ltc_res != CRYPT_OK) {
+		res = TEE_ERROR_BAD_STATE;
+		goto out;
+	}
+
+	if (mp_unsigned_bin_size(x2y2p->x) != 32 ||
+	    mp_unsigned_bin_size(x2y2p->y) != 32) {
+		res = TEE_ERROR_BAD_STATE;
+		goto out;
+	}
+
+	mp_to_unsigned_bin(x2y2p->x, x2y2);
+	mp_to_unsigned_bin(x2y2p->y, x2y2 + 32);
+
+	/* Step B4: t = KDF(x2 || y2, klen) */
+
+	/* C = C1 || C2 || C3 */
+	if (src_len <= C1_len + TEE_SM3_HASH_SIZE) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	C2_len = src_len - C1_len - TEE_SM3_HASH_SIZE;
+
+	t = calloc(1, C2_len);
+	if (!t) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	res = sm2_kdf(x2y2, sizeof(x2y2), t, C2_len);
+	if (res)
+		goto out;
+
+	if (is_zero(t, C2_len)) {
+		res = TEE_ERROR_CIPHERTEXT_INVALID;
+		goto out;
+	}
+
+	/* Step B5: get C2 from C and compute Mprime = C2 (+) t */
+
+	out_len = MIN(*dst_len, C2_len);
+	for (i = 0; i < out_len; i++)
+		dst[i] = src[C1_len + i] ^ t[i];
+	*dst_len = out_len;
+	if (out_len < C2_len) {
+		eom = calloc(1, C2_len - out_len);
+		if (!eom)
+			goto out;
+		for (i = out_len; i < C2_len; i++)
+		       eom[i - out_len] = src[C1_len + i] ^ t[i];
+	}
+
+	/* Step B6: compute u = Hash(x2 || M' || y2) and compare with C3 */
+
+	res = crypto_hash_alloc_ctx(&ctx, TEE_ALG_SM3);
+	if (res)
+		goto out;
+	res = crypto_hash_init(ctx);
+	if (res)
+		goto out;
+	res = crypto_hash_update(ctx, x2y2, 32);
+	if (res)
+		goto out;
+	res = crypto_hash_update(ctx, dst, out_len);
+	if (res)
+		goto out;
+	if (out_len < C2_len) {
+		res = crypto_hash_update(ctx, eom, C2_len - out_len);
+		if (res)
+			goto out;
+	}
+	res = crypto_hash_update(ctx, x2y2 + 32, 32);
+	if (res)
+		goto out;
+	res = crypto_hash_final(ctx, u, sizeof(u));
+	if (res)
+		goto out;
```

Comment:
We should probably use the LTC internal hash interface instead, or this can't be upstreamed.

Reply:
Problem is, the hash is SM3 which is not supported by LTC.

Reply:
OK